### PR TITLE
edit: Add del-var and del-vars

### PR DESCRIPTION
**Pull request by mmlb**
Fixes: https://github.com/elves/elvish/issues/1265
Issue: https://github.com/elves/elvish/issues/1521

### CODE REVIEW

**Comment by mmlb** [comment]
Hi @xiaq / @krader1961. This is my first working stab at getting del-var support. I also added `del-vars` for completeness. I'm not really a fan of the way I went about with the `*Deleted` methods in `eval` but that was the easiest/less disruptive change I could think of, let me know if you would like to see it done another way and I'd be happy to make the changes.

**Comment by mmlb** [comment]
My main use case for `del-var` is to combine it with direnv to avoid direnv's hook that needs to iterate through a json object calling `set-env` and `unset-env`. Also for use with nix-community/home-manager (possibly thru direnv actually).

**Comment by mmlb** [comment]
hmm although now that I look at this again, I notice that maybe `edit:add-var` and `set-env` are not equivalent, right?

**Comment by xiaq** [comment]
> I'm not really a fan of the way I went about with the `*Deleted` methods in `eval` but that was the easiest/less disruptive change I could think of, let me know if you would like to see it done another way and I'd be happy to make the changes.

I am not sure the change to `Ns` is necessary. You can just add a specialized `DeleteGlobals` method like this:

```go
func (ev *Evaler) DeleteGlobal(name string) {
  ev.mu.Lock()
  defer ev.mu.Unlock()
  g := ev.global.clone()
  for i := range g.infos {
    if g.infos[i].name == name {
      g.infos[i].deleted = true
    }
  }
  ev.global = g
}
```

And similarly for deleting multiple names.

> hmm although now that I look at this again, I notice that maybe `edit:add-var` and `set-env` are not equivalent, right?

`edit:add-var` manipulates the REPL's global namespace (variables like `$foo`) while `set-env` manipulates environment variables (like `$E:FOO`). These are completely disjoint in Elvish.

**Comment by mmlb** [comment]
> > I'm not really a fan of the way I went about with the `*Deleted` methods in `eval` but that was the easiest/less disruptive change I could think of, let me know if you would like to see it done another way and I'd be happy to make the changes.
> 
> I am not sure the change to `Ns` is necessary. You can just add a specialized `DeleteGlobals` method like this:
> 
> ```go
> func (ev *Evaler) DeleteGlobal(name string) {
>   ev.mu.Lock()
>   defer ev.mu.Unlock()
>   g := ev.global.clone()
>   for i := range g.infos {
>     if g.infos[i].name == name {
>       g.infos[i].deleted = true
>     }
>   }
>   ev.global = g
> }
> ```
> 
> And similarly for deleting multiple names.

I'll give it try with this route.

> 
> > hmm although now that I look at this again, I notice that maybe `edit:add-var` and `set-env` are not equivalent, right?
> 
> `edit:add-var` manipulates the REPL's global namespace (variables like `$foo`) while `set-env` manipulates environment variables (like `$E:FOO`). These are completely disjoint in Elvish.

Right, I was just typing-out-loud that my original reason for the PR isn't addressed by the PR :man_facepalming: . Would you be open to `edit:add-env` `edit:del-env`?

**Comment by mmlb** [comment]
@xiaq I've updated the PR with your suggestion, my edit tests passed locally so lets see if CI agrees :D.

**Comment by xiaq** [comment]
> Right, I was just typing-out-loud that my original reason for the PR isn't addressed by the PR 🤦‍♂️ . Would you be open to `edit:add-env` `edit:del-env`?

Environment variables are shared within the entire process, so you can just use `set E:ENV_NAME = ...` and `del E:ENV_NAME`.

You can also use `set-env` and `unset-env` but they are only needed when the name is dynamic.

**Comment by mmlb** [comment]
@xiaq wanted to ping you on this, even though I'm really mostly interested on `set E:ENV_NAME` and `del E:ENV_NAME`, I think its good to have the set/del pairs for this functionality.

File: pkg/edit/vars.go [hunk 1]
```diff
@@ -31,6 +34,22 @@ func addVar(fm *eval.Frame, name string, val any) error {
...
+		return errs.BadValue{
+			What:  "name argument to edit:del-var",
+			Valid: "unqualified variable name", Actual: name}
+	}
+
+	variable := fm.Evaler.Global().IndexString(name)
+	if variable == nil {
+		return errors.New("no variable $" + name)
+	}
+
```
Comment (xiaq) [inline pkg/edit/vars.go:48]: There's a race condition here when the global namespace is swapped out after the check but before the `.DeleteGlobal` call.

For dynamic deletion calls, I feel it's best to just ignore keys that don't exist. This is how `dissoc` and `del E:FOO` behave today.

File: pkg/edit/vars.go [hunk 2]
```diff
@@ -57,6 +76,32 @@ func addVars(fm *eval.Frame, m vals.Map) error {
 	return nil
 }
 
+func delVars(fm *eval.Frame, m vals.List) error {
+	for it := m.Iterator(); it.HasElem(); it.Next() {
+		n := it.Elem()
+		name, ok := n.(string)
+		if !ok {
+			return errs.BadValue{
+				What:  "name of argument to edit:del-vars",
```
Comment (xiaq) [inline pkg/edit/vars.go:85]: `name of argument to edit:del-vars` doesn't seem to be an accurate description of what the element is; `element of the argument to edit:del-vars` is better.

I think it's better to first convert the argument into a `[]string` and then delete all the arguments. Deleting them in a loop is prone to race conditions.

File: pkg/edit/vars_test.go [hunk 3]
```diff
@@ -47,3 +67,25 @@ func TestAddVars(t *testing.T) {
 		That("add-vars [&a~='']").Throws(ErrorWithType(vals.WrongType{})),
 	)
 }
+
+func TestDelVars(t *testing.T) {
+	TestWithSetup(t, func(ev *eval.Evaler) {
+		ev.ExtendGlobal(eval.BuildNs().AddGoFn("add-vars", addVars))
+		ev.ExtendGlobal(eval.BuildNs().AddGoFn("del-vars", delVars))
```
Comment (xiaq) [inline pkg/edit/vars_test.go:74]: These two `.ExtendGlobal` calls should be combined:

```go
ev.ExtendGlobal(eval.BuildNs().AddGoFns(map[string]any{"add-vars": addVars, "del-vars": delVars}))
```

File: pkg/edit/vars_test.go [hunk 4]
```diff
@@ -47,3 +67,25 @@ func TestAddVars(t *testing.T) {
...
+	},
+		That("add-vars [&foo=bar]").Then("del-vars [foo]").Then("put $foo").DoesNotCompile("variable $foo not found"),
+		That("add-vars [&a=A &b=B &c=C]").Then("del-vars [a b]").Then("put $c").Puts("C").Then("del-vars [c]").Then("put $c").DoesNotCompile("variable $c not found"),
```
Comment (xiaq) [inline pkg/edit/vars_test.go:77]: This test case seems to be testing two behaviors:

1. That deleting variables `a` and `b` leaves the variable `c` intact;
2. That deleting variable `c` makes it no longer exist.

Behavior 2 is already tested above and doesn't need to be repeated.

File: pkg/eval/eval.go [hunk 5]
```diff
@@ -174,6 +174,19 @@ func (ev *Evaler) ExtendGlobal(ns Nser) {
 	ev.global = CombineNs(ev.global, ns.Ns())
 }
 
+// DeleteGlobal deletes a binding from the global namespace.
+func (ev *Evaler) DeleteGlobal(name string) {
```
Comment (xiaq) [inline pkg/eval/eval.go:178]: Take a string slice rather than a single string.

**Pull request merged**